packages/contracts/contracts/engine/V3/EngineFactoryV0.sol

```diff
@@ -0,0 +1,215 @@
+ * The contract is initialized with an Engine and Engine Flex implementation contract, which is cloned
+ * when creating new Engine and Engine Flex Core contracts.
+ */
+contract EngineFactoryV0 is Ownable, IEngineFactoryV0 {
+    // public type
+    bytes32 public constant type_ = "EngineFactoryV0";
+
+    /**
+     * The implementation contract that is cloned when creating new Engine
+     * contracts.
+     */
+    address public immutable engineImplementation;
+    /**
+     * The implementation contract that is cloned when creating new Engine
+     * Flex contracts.
+     */
+    address public immutable engineFlexImplementation;
+
+    // Address of core registry contract.
+    address public immutable coreRegistry;
+
+    /**
+     * Indicates whether the contract is abandoned.
+     * Once abandoned, the contract can no longer be used to create new Engine
+     * and Engine Flex contracts.
+     */
+    bool public isAbandoned; // default false
+
+    /**
+     * @notice validates and assigns immutable configuration variables
+     * @param engineImplementation_ address of the Engine
+     * implementation contract
+     * @param engineFlexImplementation_ address of the Engine Flex
+     * implementation contract
+     * @param coreRegistry_ address of the core registry contract
+     */
+    constructor(
+        address engineImplementation_,
+        address engineFlexImplementation_,
+        address coreRegistry_
+    ) {
+        coreRegistry = coreRegistry_;
+        engineFlexImplementation = engineFlexImplementation_;
+        engineImplementation = engineImplementation_;
+        // set owner of core registry contract
+        Ownable(coreRegistry_).transferOwnership(address(this));
+        // emit event
+        emit Deployed({
+            engineImplementation: engineImplementation_,
+            engineFlexImplementation: engineFlexImplementation_,
+            type_: type_
+        });
+    }
+
+    /**
+     * @notice Creates a new Engine or Engine Flex contract with the provided
+     * `engineConfiguration`, depending on the `engineCoreType`.
+     * @param engineCoreType Type of Engine Core contract.
+     * @param engineConfiguration EngineConfiguration data to configure the
+     * contract with.
+     * @param adminACLContract Address of admin access control contract, to be
+     * set as contract owner. A new contract will be deployed if address is null.
+     * @return engineContract The address of the newly created Engine or Engine Flex
+     * contract. The address is also emitted in both the `EngineCreated` and
+     * `EngineFlexCreated` events.
+     */
+    function createEngineContract(
+        IEngineFactoryV0.EngineCoreType engineCoreType,
+        EngineConfiguration calldata engineConfiguration,
+        address adminACLContract
+    ) external onlyOwner returns (address engineContract) {
+        require(!isAbandoned, "factory is abandoned");
+        // validate engine contract configuration
```

Comment:
would it make sense to validate that adminACLContract and newSuperAdminAddress are either both zero or both non-zero?

Reply:
Talking through this one out loud, LMK if this is correct 🤔 : 

I think one of two things can happen, either:

1. There is no existing AdminACL Contract so we deploy a new one and transfer ownership to the newSuperAdminAddress.. In this case the passed in `adminACLContract` would be the null address (indicating we need to deploy a new one) and the passed in newSuperAdminAddress would be "0x..."

2. There is an existing AdminACL Contract, passed in as "0x...", which we use to initialize and transfer ownership of the new Engine contract (no need to deploy a new one but in this case whether the newSuperAdmin address was null or not, we can't call transferOwnership of the AdminACL contract bc it would need to be called by the super admin)

So I think the only check we would need to make is: 
`if (adminACLContract == address(0)) -> newSuperAdminAddress != address(0)`

Reply:
ah, my bad, yes you are correct! do you think it also makes sense to also validate that they don't pass in an existing AdminACL contract with a non-null newSuperAdmin, because that wouldn't be possible to assign in practice?

Reply:
that's a good call! is there any other validation that should be done there? validating that the newSuperAdmin address passed in is the owner of the AdminACL contract passed in?

Reply:
lol sorry I think not helping 🤣 wdyt about the following logic?
```
if (adminACLContract == addr(0)) {
    // require non-zero newSuperAdminAddress
    // deploy new splitter, assign to new super admin
} else {
    // require zero newSuperAdminAddress
    // use existing splitter, a non-zero new super admin is nonsense here
}
```

Reply:
or could totally do what you describe and just check the superAdmin on an existing superAdmin 👍 I think either way would be good!